File: src/city.py
#!/usr/bin/env python
# encoding: utf-8

if __name__=='__main__':
    city_dict = {}
    city_vect = []

    with open("dat/city_pop.tsv", "r") as f:
        count = 0

        for line in f:
            (city, population) = line.strip().split("\t")

            if (count > 0) and (city not in city_dict):
                city_dict[city] = int(population)

            count += 1

    with open("dat/city_geo.tsv", "r") as f:
        count = 0

        for line in f:
            (city, latitude, longitude) = line.strip().split("\t")

            if (count > 0) and (city in city_dict):
                city_vect.append([city_dict[city], city, float(latitude), float(longitude)])

            count += 1

    city_vect.sort(reverse=True)
    sum_pop = float(sum([ v[0] for v in city_vect]))
    cum_prob = 0.0

    for (pop, city, lat, lng) in city_vect:
        cum_prob += pop/sum_pop
        print "%0.6f\t%s\t%0.4f\t%0.4f" % (cum_prob, city, lat, lng)

File: scala/city.scala
val city_pop = sc.textFile("dat/city_pop.tsv").map(_.split("\t")).map(r => (r(0), r))
val city_geo = sc.textFile("dat/city_geo.tsv").map(_.split("\t")).map(r => (r(0), r))

case class City (pop: Float, name: String, lat: Float, lng: Float)

val city_vect = city_pop.join(city_geo).map( 
 r => City(r._2._1(1).toFloat, r._1, r._2._2(1).toFloat, r._2._2(2).toFloat)
).map( r => (r.pop, r) ).sortByKey(false)

val sum_pop = city_vect.map(r => r._1).sum()
var cum_prob = 0.0

val cities = city_vect.map( r => r._2 ).collect

def cdf (c: City): String = {
  cum_prob += c.pop / sum_pop
  "%f\t%s\t%f\t%f".format(cum_prob, c.name, c.lat, c.lng)
}

sc.makeRDD( for (c <- cities) yield cdf(c) ).saveAsTextFile("city_prob")
